### PR TITLE
Support SSMS scripting for Sequences

### PR DESCRIPTION
**Pull request by hash-16**
### Description

Earlier we were not able to script due to absence of sys.sequences view . Added the sys.sequences view to support scripting.

Signed-off-by: Ashish Prasad [pashisht@amazon.com](mailto:pashisht@amazon.com)

### Issues Resolved
BABEL-3920

### Test Scenarios Covered ###
* **Use case based -**
NA

* **Boundary conditions -**
NA

* **Arbitrary inputs -**
NA

* **Negative test cases -**
NA

* **Minor version upgrade tests -**
Added

* **Major version upgrade tests -**
Added

* **Performance tests -**
NA

* **Tooling impact -**
Tested

* **Client tests -**
Tested

### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: test/python/input/ddl_sequence.sql [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+Create schema Test_Seq
+GO
+
+CREATE SEQUENCE Test_Seq.test START WITH 1 INCREMENT BY 1 ;
+GO
+
+CREATE SEQUENCE Test_Seq.test1 START WITH 5 INCREMENT BY 5 ;  
+GO
+
+CREATE SEQUENCE Test_Seq.test2 START WITH 24329 INCREMENT BY 1 ;
```
Comment (shalinilohia50) [inline test/python/input/ddl_sequence.sql:41]: Can you please add more tests covering scenarios like AS clause, CYCLE keyword, minvalue, maxvalue etc.?

File: contrib/babelfishpg_tsql/sql/sys_views.sql [hunk 2]
```diff
@@ -3004,3 +3004,39 @@ SELECT
...
+  , CAST(s.schema_id as int) as schema_id
+  , CAST(0 as int) as parent_object_id
+  , CAST('SO' as char(2)) as type
+  , CAST('SEQUENCE_OBJECT' as sys.nvarchar(60)) as type_desc
+  , CAST(null as sys.datetime) as create_date
+  , CAST(null as sys.datetime) as modify_date
+  , CAST(0 as sys.bit) as is_ms_shipped
+  , CAST(0 as sys.bit) as is_published
+  , CAST(0 as sys.bit) as is_schema_published
+  , CAST(ps.seqstart as sql_variant ) as start_value
```
Comment (shalinilohia50) [inline contrib/babelfishpg_tsql/sql/sys_views.sql]: sys.sql_variant

File: contrib/babelfishpg_tsql/sql/sys_views.sql [hunk 3]
```diff
@@ -3004,3 +3004,39 @@ SELECT
...
+  , CASE ps.seqcycle when 't' then CAST(1 as sys.bit) else CAST(0 as sys.bit) end as is_cycling
+  , CAST(ps.seqcache as sys.bit ) as is_cached
+  , CAST(NULL as int ) as cache_size
+  , CAST(ps.seqtypid as int ) as system_type_id
+  , CAST(ps.seqtypid as int ) as user_type_id
+  , CAST(0 as sys.tinyint ) as precision
+  , CAST(0 as sys.tinyint ) as scale
+  , CAST('ABC' as sql_variant  ) as current_value
+  , CAST(0 as sys.bit ) as is_exhausted
+  , CAST('ABC' as sql_variant ) as last_used_value
```
Comment (shalinilohia50) [inline contrib/babelfishpg_tsql/sql/sys_views.sql]: it can be `ps.last_value`

File: contrib/babelfishpg_tsql/sql/sys_views.sql [hunk 4]
```diff
@@ -3004,3 +3004,39 @@ SELECT
...
+  , CAST(0 as sys.bit) as is_ms_shipped
+  , CAST(0 as sys.bit) as is_published
+  , CAST(0 as sys.bit) as is_schema_published
+  , CAST(ps.seqstart as sql_variant ) as start_value
+  , CAST(ps.seqincrement as sql_variant ) as increment
+  , CAST(ps.seqmin as sql_variant  ) as minimum_value
+  , CAST(ps.seqmax as sql_variant ) as maximum_value
+  , CASE ps.seqcycle when 't' then CAST(1 as sys.bit) else CAST(0 as sys.bit) end as is_cycling
+  , CAST(ps.seqcache as sys.bit ) as is_cached
+  , CAST(NULL as int ) as cache_size
```
Comment (shalinilohia50) [inline contrib/babelfishpg_tsql/sql/sys_views.sql]: Can be `ps.cache_size`

File: contrib/babelfishpg_tsql/sql/sys_views.sql [hunk 5]
```diff
@@ -3004,3 +3004,39 @@ SELECT
     CAST(0 as sys.BIT) AS is_ambiguous
 WHERE FALSE;
 GRANT SELECT ON sys.sql_expression_dependencies TO PUBLIC;
+
```
Comment (shalinilohia50) [inline contrib/babelfishpg_tsql/sql/sys_views.sql:3054]: Nit: Extra line.

File: contrib/babelfishpg_tsql/sql/sys_views.sql [hunk 6]
```diff
@@ -3004,3 +3004,39 @@ SELECT
...
+  , CAST('SO' as char(2)) as type
+  , CAST('SEQUENCE_OBJECT' as sys.nvarchar(60)) as type_desc
+  , CAST(null as sys.datetime) as create_date
+  , CAST(null as sys.datetime) as modify_date
+  , CAST(0 as sys.bit) as is_ms_shipped
+  , CAST(0 as sys.bit) as is_published
+  , CAST(0 as sys.bit) as is_schema_published
+  , CAST(ps.seqstart as sql_variant ) as start_value
+  , CAST(ps.seqincrement as sql_variant ) as increment
+  , CAST(ps.seqmin as sql_variant  ) as minimum_value
```
Comment (shalinilohia50) [inline contrib/babelfishpg_tsql/sql/sys_views.sql]: `sys.sql_variant` applies at all the places.

**Review by shalinilohia50 — CHANGES_REQUESTED** [review]

**Comment by shalinilohia50** [comment]
Please resolve merge conflicts.

File: contrib/babelfishpg_tsql/sql/sys_views.sql (hunk 3 above)
Comment (hash-16) [reply contrib/babelfishpg_tsql/sql/sys_views.sql]: pg_sequence does not have column last_value

Comment (shalinilohia50) [reply contrib/babelfishpg_tsql/sql/sys_views.sql]: Official doc for PG suggests so. https://www.postgresql.org/docs/15/view-pg-sequences.html

Comment (shalinilohia50) [reply contrib/babelfishpg_tsql/sql/sys_views.sql]: Oh! It is _pg_sequences_, not _pg_sequence_. Can we use _pg_sequences_ instead?

Comment (hash-16) [reply contrib/babelfishpg_tsql/sql/sys_views.sql]: I think we cannot use directly as it does not a primary key to do the join.

File: test/python/input/ddl_sequence.sql (hunk 1 above)
Comment (hash-16) [reply test/python/input/ddl_sequence.sql:41]: Added

File: contrib/babelfishpg_tsql/sql/sys_views.sql (hunk 4 above)
Comment (hash-16) [reply contrib/babelfishpg_tsql/sql/sys_views.sql]: Done

File: contrib/babelfishpg_tsql/sql/sys_views.sql (hunk 2 above)
Comment (hash-16) [reply contrib/babelfishpg_tsql/sql/sys_views.sql]: Done

File: contrib/babelfishpg_tsql/sql/sys_views.sql (hunk 6 above)
Comment (hash-16) [reply contrib/babelfishpg_tsql/sql/sys_views.sql]: Done

File: contrib/babelfishpg_tsql/sql/sys_views.sql (hunk 5 above)
Comment (hash-16) [reply contrib/babelfishpg_tsql/sql/sys_views.sql:3054]: Done

File: contrib/babelfishpg_tsql/sql/sys_views.sql [hunk 7]
```diff
@@ -3067,3 +3103,4 @@ SELECT
     CAST('' as sys.NVARCHAR(60)) AS state_desc
 WHERE FALSE;
 GRANT SELECT ON sys.database_permissions TO PUBLIC;
+
```
Comment (shalinilohia50) [inline contrib/babelfishpg_tsql/sql/sys_views.sql]: Nit: unnecessary line.

Comment (hash-16) [reply contrib/babelfishpg_tsql/sql/sys_views.sql]: Done

File: contrib/babelfishpg_tsql/sql/sys_views.sql [hunk 8]
```diff
@@ -3052,6 +3052,42 @@ SELECT
...
 
+
+create or replace view sys.sequences as
+select
+    CAST(p.relname as sys.sysname) as name
+  , CAST(p.oid as int) as object_id
+  , CAST(null as int) as principal_id
+  , CAST(s.schema_id as int) as schema_id
+  , CAST(0 as int) as parent_object_id
+  , CAST('SO' as char(2)) as type
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/sql/sys_views.sql]: sys.bpchar

File: test/JDBC/expected/sys_sequences-vu-prepare.out [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+BEGIN
+    RETURN (SELECT COUNT(*) FROM sys.sequences WHERE is_cycling= 0)
+END
+GO
+
+CREATE FUNCTION sys_sequences_vu_prepare_func1()
+RETURNS INT
+AS
+BEGIN
+    RETURN (SELECT COUNT(*) FROM sys.sequences WHERE name='Babel_DATAtype_SQLvariant_vu_prepare_t4_SEC');
```
Comment (Deepesh125) [inline test/JDBC/expected/sys_sequences-vu-prepare.out]: who is creating this sequence?

**Review by Deepesh125 — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/sql/sys_views.sql (hunk 8 above)
Comment (hash-16) [reply contrib/babelfishpg_tsql/sql/sys_views.sql]: Done

File: test/JDBC/expected/sys_sequences-vu-verify.out [hunk 10]
```diff
@@ -0,0 +1,44 @@
+USE master
+GO
+
+SELECT * FROM sys_sequences_vu_prepare_view
+GO
+~~START~~
+varchar#!#int#!#int#!#char#!#nvarchar#!#datetime#!#datetime#!#bit#!#bit#!#bit#!#sql_variant#!#sql_variant#!#sql_variant#!#sql_variant#!#bit#!#bit#!#int#!#int#!#int#!#tinyint#!#tinyint#!#sql_variant#!#bit#!#sql_variant
+test_seq#!#<NULL>#!#0#!#SO#!#SEQUENCE_OBJECT#!#<NULL>#!#<NULL>#!#0#!#0#!#0#!#1#!#1#!#1#!#2147483647#!#0#!#0#!#1#!#23#!#23#!#0#!#0#!#ABC#!#0#!#ABC
+?????????????????#!#<NULL>#!#0#!#SO#!#SEQUENCE_OBJECT#!#<NULL>#!#<NULL>#!#0#!#0#!#0#!#1#!#1#!#1#!#9223372036854775807#!#0#!#0#!#1#!#20#!#20#!#0#!#0#!#ABC#!#0#!#ABC
```
Comment (Deepesh125) [inline test/JDBC/expected/sys_sequences-vu-verify.out]: in babelfish, sysname is defined as sys.varchar(128) which should have been sys.nvarchar(128) (as per T-SQL doc) and that is the reason we are seeing `?` here. Would it be possible to use sys.nvarchar(128) for `name` column?

Comment (hash-16) [reply test/JDBC/expected/sys_sequences-vu-verify.out]: Addressed.

**Review by Deepesh125 — APPROVED** [review]

**Pull request merged**